src/css.js

```diff
@@ -336,16 +337,25 @@ jQuery.each( [ "height", "width" ], function( i, name ) {
 		},
 
 		set: function( elem, value, extra ) {
-			var styles = extra && getStyles( elem );
-			return setPositiveNumber( elem, value, extra ?
-				augmentWidthOrHeight(
+			var matches,
+				styles = extra && getStyles( elem ),
+				subtract = extra && augmentWidthOrHeight(
 					elem,
 					name,
 					extra,
 					jQuery.css( elem, "boxSizing", false, styles ) === "border-box",
 					styles
-				) : 0
-			);
+				);
+
+			// Convert to pixels if value adjustment is needed
+			if ( subtract && ( matches = rcssNum.exec( value ) ) &&
+				( matches[ 3 ] || "px" ) !== "px" ) {
```

Comment:
Does this need a `!matches[ 1 ]` guard? Without it, `.width( "+=10%" )` will:
* `elem.style[ "width" ] = "+=10%"` (noop, or are there still supported environments where this throws?)
* `value = jQuery.css( elem, "width" )` (current width of {X}px)
* `return setPositiveNumber( elem, "{X}px", subtract )`
  * `var matches = rcssNum.exec( value )` (`[ "{X}px", undefined, X, "px" ]`)
  * `return Math.max( 0, X - ( subtract || 0 ) ) + "px" )`
* Set inline `width` to that value

Whereas with a guard, we'd just set an inline width of the obviously invalid `+=10%`.